src/com/google/enterprise/adaptor/opentext/OpentextAdaptor.java

```diff
@@ -657,134 +657,180 @@ public void getDocIds(DocIdPusher pusher) throws InterruptedException {
     if (!this.markAllDocsAsPublic) {
       MemberService memberService =
           this.soapFactory.newMemberService(documentManagement);
-      Map<GroupPrincipal, List<Principal>> groupDefinitions =
-          getGroups(memberService);
-      if (this.publicAccessGroupEnabled) {
-        List<Principal> publicAccessGroup = getPublicAccessGroup(memberService);
-        if (publicAccessGroup.size() > 0) {
-          groupDefinitions.put(
-              new GroupPrincipal("[Public Access]", this.localNamespace),
-              publicAccessGroup);
+      DocIdPusher.FeedType feedType = DocIdPusher.FeedType.FULL;
+      SOAPFaultException groupsException = null;
+      Groups groups = new Groups(memberService, this);
+      try {
+        groups.addGroups();
+        if (this.publicAccessGroupEnabled) {
+          groups.addPublicAccessGroup("[Public Access]", this.localNamespace);
         }
+        if (!groups.isFull()) {
+          feedType = DocIdPusher.FeedType.INCREMENTAL;
+        }
+      } catch (SOAPFaultException e) {
+        groupsException = e;
+        feedType = DocIdPusher.FeedType.INCREMENTAL;
+        log.log(Level.WARNING, "Error while reading groups", e);
       }
+
+      // If some groups were found, push them; otherwise, if an
+      // exception was thrown, rethrow it as before.
+      Map<GroupPrincipal, List<Principal>> groupDefinitions =
+          groups.getGroupDefinitions();
       if (groupDefinitions.size() > 0) {
-        pusher.pushGroupDefinitions(groupDefinitions, true);
+        if (feedType == DocIdPusher.FeedType.INCREMENTAL) {
+          log.log(Level.FINE, "Sending partial groups results");
+        }
+        pusher.pushGroupDefinitions(groupDefinitions,
+            /* case sensitive */ true, feedType,
+            /* group source */ null, /* exception handler */ null);
+      } else if (groupsException != null) {
```

Comment:
Is it possible for there to be a groupsException and goupDefintiions.size() to be 0?   Should the code first check groupsException and then groupsDefirtinitions.size()?

Reply:
I'm pretty sure it's possible, but we have generally preferred to send partial results as INCREMENTAL, rather than not send anything at all when we have some results and then get an exception. (We can at least update the group definitions that we did get, for example.)